### PR TITLE
Fix and extend BinomialHeap

### LINKED ISSUE #629: BinomialHeap does not work when multiple elements have the same priority
MWE:

```lean
import Std.Data.BinomialHeap

open Std

def test : BinomialHeap (Nat × Nat) (λ (x, _) (y, _) => x < y) :=
  BinomialHeap.empty.insert (0, 1) |>.insert (0, 2) |>.insert (0, 3)

#eval test.toList
```

Expected output of the `#eval`: `[(0, 1), (0, 2), (0, 3)]` (or a permutation).
Actual output: `[(0, 2), (0, 2), (0, 1)]`.

Lean version: current master.

### PR DESCRIPTION
**Pull request by JLimperg**
Major changes:

* remove redundant constructor of `Heap`
* fix #629: `head` and `tail` now agree even if the comparison function is not irreflexive
* add `ofList`, `ofArray`, `deleteMin`, `tail?`, `toArray`, `toListUnordered`, `toArrayUnordered`
* add a simple test case

Each commit is self-contained, so you can review them individually. I can also split the PR if that's more convenient for you.

### CODE REVIEW

File: src/Std/Data/BinomialHeap.lean [hunk 1]
```diff
@@ -26,122 +32,210 @@ def hRank : List (HeapNode α) → Nat
...
+      | heap ns => do
+        let mut acc := acc
+        for n in ns do
+          acc := acc.push n.val
+          for h in n.children do
+            acc := go acc h
+        return acc
+
+inductive WellFormed (le : α → α → Bool) : Heap α → Prop where
+  | emptyWff                : WellFormed le empty
```
[the remark this answers is not in the document]
Comment (JLimperg) [reply src/Std/Data/BinomialHeap.lean]: I've renamed the constructors to `empty` etc. Is there a fixed naming convention for theorems now? I recall some discussion about capitalisation a while ago.

File: tests/lean/629.lean [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import Std.Data.BinomialHeap
+
+open Std
+
+def test : BinomialHeap (Nat × Nat) (λ (x, _) (y, _) => x < y) :=
```
[the remark this answers is not in the document]
Comment (JLimperg) [reply tests/lean/629.lean:5]: The data structure should work with both lt and le (now). I only renamed lt to le in the implementation as a reminder that we should not implicitly rely on irreflexivity. I'd be happy to use the le typeclass if there's consensus on this.

File: src/Std/Data/BinomialHeap.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kha) [reply src/Std/Data/BinomialHeap.lean]: There is, https://github.com/leanprover/lean4/issues/402

[the remark this answers is not in the document]
Comment (JLimperg) [reply src/Std/Data/BinomialHeap.lean]: Ah, good to know. Thanks!

**Comment by Kha** [comment]
> These are implementation notes rather than user-level comments, that's why I didn't make them doc comments

I agree with this fwiw, these would be *terrible* docstrings :sweat_smile: .

**Pull request merged**